### Exercise 2: House prices model

- Load the houseprices data from Thinkful's database.
- Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC.
- Do you think your model is satisfactory? If so, why?
- In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables.
- For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import statsmodels.api as sm

import warnings
warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
house_df = pd.read_sql_query('select * from houseprices',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

In [6]:
house_df = pd.concat([house_df,pd.get_dummies(house_df.mszoning, prefix="mszoning", drop_first=True)], axis=1)
house_df = pd.concat([house_df,pd.get_dummies(house_df.street, prefix="street", drop_first=True)], axis=1)
dummy_column_names = list(pd.get_dummies(house_df.mszoning, prefix="mszoning", drop_first=True).columns)
dummy_column_names = dummy_column_names + list(pd.get_dummies(house_df.street, prefix="street", drop_first=True).columns)

In [7]:
X = house_df[['overallqual', 'grlivarea', 'garagecars', 'garagearea', 'totalbsmtsf', 'mszoning_FV', 'mszoning_RL']]
Y = house_df.saleprice

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.769
Model:                            OLS   Adj. R-squared:                  0.768
Method:                 Least Squares   F-statistic:                     689.5
Date:                Mon, 05 Aug 2019   Prob (F-statistic):               0.00
Time:                        21:37:31   Log-Likelihood:                -17475.
No. Observations:                1460   AIC:                         3.497e+04
Df Residuals:                    1452   BIC:                         3.501e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.085e+05   4791.089    -22.653      0.000   -1.18e+05   -9.91e+04
overallqual  2.341e+04   1079.265     21.690      0.000    2.13e+04    2.55e+04
grlivarea      45.6643      2.459     18.569      0.000      40.841      50.488
garagecars   1.347e+04   2983.458      4.513      0.000    7613.379    1.93e+04
garagearea     15.9352     10.345      1.540      0.124      -4.358      36.229
totalbsmtsf    28.3424      2.927      9.682      0.000      22.600      34.085
mszoning_FV  1.607e+04   5577.883      2.882      0.004    5131.939     2.7e+04
mszoning_RL   1.96e+04   2832.590      6.919      0.000     1.4e+04    2.52e+04
==============================================================================
Omnibus:                      415.658   Durbin-Watson:                   1.980
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            41074.355
Skew:                          -0.117   Prob(JB):                         0.00
Kurtosis:                      28.983   Cond. No.                     1.15e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.15e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- F-test = 689.5
- R-squared = 0.769
- adjusted R-squared = 0.768
- AIC = 34,970
- BIC = 35,010

The p value associated with the F-test is zero, which is statistically significant and means our model is useful. Then looking at the R-squared and adjusted R-squared values, we note that they are relatively high (but not too high as to suggest over-fitting), so this is a good indication as well that our model is well fitted. For a sample size of 1460, the AIC and BIC are also realtively low, but they are most useful for comparisons. I believe this model is satisfactory given that the p-value is so low and the R-squared value is high enough that this model will have significant predictive power.

In [11]:
X = house_df[['overallqual', 'grlivarea', 'garagecars',  'totalbsmtsf', 'mszoning_FV', 'mszoning_RL']]
Y = house_df.saleprice

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.768
Model:                            OLS   Adj. R-squared:                  0.767
Method:                 Least Squares   F-statistic:                     803.3
Date:                Mon, 05 Aug 2019   Prob (F-statistic):               0.00
Time:                        21:46:13   Log-Likelihood:                -17476.
No. Observations:                1460   AIC:                         3.497e+04
Df Residuals:                    1453   BIC:                         3.500e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.087e+05   4792.537    -22.674      0.000   -1.18e+05   -9.93e+04
overallqual  2.332e+04   1078.278     21.628      0.000    2.12e+04    2.54e+04
grlivarea      45.9659      2.452     18.743      0.000      41.155      50.777
garagecars   1.721e+04   1732.720      9.931      0.000    1.38e+04    2.06e+04
totalbsmtsf    29.3276      2.858     10.261      0.000      23.721      34.934
mszoning_FV  1.679e+04   5560.901      3.020      0.003    5885.005    2.77e+04
mszoning_RL  1.968e+04   2833.445      6.946      0.000    1.41e+04    2.52e+04
==============================================================================
Omnibus:                      406.420   Durbin-Watson:                   1.977
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            37605.498
Skew:                          -0.057   Prob(JB):                         0.00
Kurtosis:                      27.863   Cond. No.                     1.11e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.11e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

This model where we have taken out the feature **garagearea** is almost exactly the same as the model with the feature included, so it is reasonable that the model would be better without it because the fewer the features, the better. When we look at the p-value of garagearea compared to the p-values of the other features, we see that it is significantly higher, so it would make sense that garagearea would give the model the least amount of predictive power. We also noted before that garagearea and garagecars are highly correlated, so it would make sense that we would only need one of the two features to have the same amount of predictive power in our model.

In [12]:
house_df['overallqual_grlivarea_interaction'] = house_df.overallqual * house_df.grlivarea
X = house_df[['overallqual', 'grlivarea', 'garagecars', 'totalbsmtsf', 'mszoning_FV', 'mszoning_RL', 'overallqual_grlivarea_interaction']]
Y = house_df.saleprice

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.785
Model:                            OLS   Adj. R-squared:                  0.784
Method:                 Least Squares   F-statistic:                     758.0
Date:                Mon, 05 Aug 2019   Prob (F-statistic):               0.00
Time:                        21:56:12   Log-Likelihood:                -17421.
No. Observations:                1460   AIC:                         3.486e+04
Df Residuals:                    1452   BIC:                         3.490e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
=====================================================================================================
                                        coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------
const                              -717.0561   1.11e+04     -0.064      0.949   -2.26e+04    2.11e+04
overallqual                        6970.7136   1853.975      3.760      0.000    3333.958    1.06e+04
grlivarea                           -28.5873      7.390     -3.869      0.000     -43.083     -14.092
garagecars                         1.848e+04   1673.643     11.042      0.000    1.52e+04    2.18e+04
totalbsmtsf                          23.9094      2.800      8.538      0.000      18.416      29.402
mszoning_FV                         2.04e+04   5368.243      3.799      0.000    9864.706    3.09e+04
mszoning_RL                        2.148e+04   2735.084      7.854      0.000    1.61e+04    2.68e+04
overallqual_grlivarea_interaction    11.0255      1.035     10.648      0.000       8.994      13.057
==============================================================================
Omnibus:                     1099.515   Durbin-Watson:                   2.009
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           187982.423
Skew:                          -2.588   Prob(JB):                         0.00
Kurtosis:                      58.347   Cond. No.                     1.30e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.3e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

This model includes the interaction between the features **overallqual** and **grlivarea** as its 7th feature. We see that the adjusted R-squared value 0.784, whereas it was 0.767 before, so this is a good improvement over the previous model. 

In [13]:
house_df['garagecars_grlivarea_interaction'] = house_df.garagecars * house_df.grlivarea
X = house_df[['overallqual', 'grlivarea', 'garagecars', 'totalbsmtsf', 'mszoning_FV', 'mszoning_RL', 'garagecars_grlivarea_interaction']]
Y = house_df.saleprice

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.796
Model:                            OLS   Adj. R-squared:                  0.795
Method:                 Least Squares   F-statistic:                     807.7
Date:                Mon, 05 Aug 2019   Prob (F-statistic):               0.00
Time:                        22:02:21   Log-Likelihood:                -17385.
No. Observations:                1460   AIC:                         3.479e+04
Df Residuals:                    1452   BIC:                         3.483e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
====================================================================================================
                                       coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------
const                            -2.366e+04   7583.493     -3.119      0.002   -3.85e+04   -8779.878
overallqual                       2.254e+04   1014.609     22.217      0.000    2.06e+04    2.45e+04
grlivarea                          -12.7702      4.805     -2.658      0.008     -22.195      -3.345
garagecars                       -2.605e+04   3506.099     -7.431      0.000   -3.29e+04   -1.92e+04
totalbsmtsf                         27.7188      2.688     10.313      0.000      22.447      32.991
mszoning_FV                       2.177e+04   5236.778      4.157      0.000    1.15e+04     3.2e+04
mszoning_RL                       2.074e+04   2663.168      7.788      0.000    1.55e+04     2.6e+04
garagecars_grlivarea_interaction    30.0021      2.154     13.931      0.000      25.778      34.227
==============================================================================
Omnibus:                      712.362   Durbin-Watson:                   1.960
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            61446.048
Skew:                          -1.357   Prob(JB):                         0.00
Kurtosis:                      34.666   Cond. No.                     3.31e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.31e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

This model includes the interaction between the features **garagecars** and **grlivarea** as its 7th feature. We see that the adjusted R-squared value 0.795, whereas it was 0.784 and previously 0.767 before, so this is a good improvement over both of the previous models. I think if we were to continue adding the interactions of all the features one by one, we would eventually come up with the interaction with the best predictive power, but 0.795 is pretty good by itself, so this isn't entirely necessary.